### PR TITLE
[NIC-3] Add sign in and sign up logic

### PR DESCRIPTION
**Pull request by AksyonovaKaterina**

### CODE REVIEW

**Review by alelambin — APPROVED** [review]

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/NotFoundException.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package com.nosql.authservice.common.exception
+
+import com.nosql.authservice.common.exception.enumerator.ApplicationExceptionCodeEnum.NOT_FOUND_ERROR
+import org.springframework.http.HttpStatus
+
+open class NotFoundException(
+    status: HttpStatus = HttpStatus.NOT_FOUND,
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/NotFoundException.kt]: maybe use `HttpStatus` only in the `ApplicationException` constructor?

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    @ExceptionHandler(WebExchangeBindException::class)
+    fun handleException(exception: WebExchangeBindException): ResponseEntity<ApplicationErrorDto>? {
+
+        val errorMessageJoiner = StringJoiner(ERROR_MESSAGE_DELIMITER)
+        val bindingResult = exception.bindingResult
+
+        if (bindingResult.hasGlobalErrors()) {
+
+            val globalErrorsMessages = bindingResult.globalErrors.stream()
+                .map { String.format("%s: %s", it.objectName, it.defaultMessage) }
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]: You could use kotlin string template

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    private val log: Logger by logger()
+
+    @ExceptionHandler(ApplicationException::class)
+    fun handleCommonException(exception: ApplicationException): ResponseEntity<ApplicationErrorDto> {
+
+        val errorDto = ApplicationErrorDto(exception)
+        val status = exception.status
+
+        log.error("Got application error. Response: $errorDto")
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]:
```suggestion
        log.error("Got application exception. Response: $errorDto")
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+            errorMessageJoiner.add(fieldErrorMessages)
+        }
+
+        val errorDto = ApplicationErrorDto(
+            code = VALIDATION_ERROR.name,
+            description = errorMessageJoiner.toString(),
+        )
+
+        log.error("Got validation error. Response: $errorDto", exception)
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]:
```suggestion
        log.error("Got validation exception. Response: $errorDto", exception)
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            .body(errorDto)
+    }
+
+    @ExceptionHandler(ServerWebInputException::class)
+    fun handleException(exception: ServerWebInputException): ResponseEntity<ApplicationErrorDto> {
+        val errorDto = ApplicationErrorDto(
+            code = DECODING_ERROR.name,
+            description = getBadlyDecodedFieldMessage(exception.cause?.message)
+        )
+        log.warn("Got decoding error. Response: {}", errorDto, exception)
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]:
```suggestion
        log.warn("Got decoding exception. Response: {}", errorDto, exception)
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            .badRequest()
+            .body(errorDto)
+    }
+
+    @ExceptionHandler(Exception::class)
+    fun handleUnexpectedException(exception: Exception): ResponseEntity<ApplicationErrorDto> {
+
+        val apiErrorDto = ApplicationErrorDto(exception)
+
+        log.error("Got unexpected application error. Response: {}", apiErrorDto, exception)
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt:90]:
```suggestion
        log.error("Got unexpected application exception. Response: {}", apiErrorDto, exception)
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/UserComponent.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package com.nosql.authservice.component
+
+import com.nosql.authservice.dto.UserDto
+import com.nosql.authservice.entity.UserEntity
+
+interface UserComponent {
+
+    suspend fun saveIfNotExists(user: UserEntity): UserEntity
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/UserComponent.kt:8]:
```suggestion
    suspend fun save(user: UserEntity): UserEntity
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        logger.logBefore(operationDetails)
+
+        return userRepository.existsByLogin(login)
+            .onErrorMap { handleError(it, operationDetails) }
+            .doOnSuccess { logger.logSuccess(operationDetails) }
+            .awaitSingle()
+    }
+
+    @Transactional(readOnly = true)
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt:55]: You could remove the `@Transactional` annotation from the class and add it only for write methods.

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+import org.springframework.transaction.annotation.Transactional
+import reactor.core.publisher.Mono
+
+@Component
+@Transactional
+class DefaultUserComponent(
+    private val userRepository: UserRepository,
+) : UserComponent {
+
+    private val logger: Logger by logger()
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: I think you could simplify it:
```suggestion
    private val log: Logger by logger()
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Transactional
+    override suspend fun saveIfNotExists(user: UserEntity): UserEntity {
+        val operationDetails = "Save if not exists 'auth' record with login = "
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: login?

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        return userRepository.existsByLogin(login)
+            .onErrorMap { handleError(it, operationDetails) }
+            .doOnSuccess { logger.logSuccess(operationDetails) }
+            .awaitSingle()
+    }
+
+    @Transactional(readOnly = true)
+    override suspend fun getByLoginAndPasswordHash(userDto: UserDto): UserEntity {
+        val (login, passwordHash) = userDto
+        val operationDetails = "Get 'auth' record by login = '$login' and password_hash = '$passwordHash'"
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: I think you should not log the password hash anywhere

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+        logger.logBefore(operationDetails)
+
+        return userRepository.getByLoginAndPasswordHash(login, passwordHash)
+            .switchIfEmpty(
+                Mono.error(
+                    NotFoundException(
+                        description = "Not found 'auth' record by given login = '$login' and " +
+                                "password_hash = '$passwordHash'"
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: password hash

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    @Transactional(readOnly = true)
+    override suspend fun getByLoginAndPasswordHash(userDto: UserDto): UserEntity {
+        val (login, passwordHash) = userDto
+        val operationDetails = "Get 'auth' record by login = '$login' and password_hash = '$passwordHash'"
+
+        logger.logBefore(operationDetails)
+
+        return userRepository.getByLoginAndPasswordHash(login, passwordHash)
+            .switchIfEmpty(
+                Mono.error(
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: You could use `NotFoundException(...).toMono()` instead

**Review by V7074 — CHANGES_REQUESTED** [review]

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/NotFoundException.kt (hunk 1 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/NotFoundException.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+        val operationFailedWithMessage = "$operationDetails: failed with"
+        val errorDetails = if (error is ApplicationException) error.description else "something went wrong"
+
+        return when (error) {
+            is DataIntegrityViolationException -> {
+                DataIntegrityViolationApplicationException()
+            }
+            is ApplicationException -> {
+                error.apply { description = """$operationFailedWithMessage error: $errorDetails""" }
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]:
```suggestion
                error.apply { description = "$operationFailedWithMessage error: $errorDetails" }
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+            }
+            else -> {
+                ApplicationException(
+                    code = ApplicationExceptionCodeEnum.UNEXPECTED_ERROR,
+                    description = """$operationFailedWithMessage unexpected error: $errorDetails""",
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]:
```suggestion
                    description = "$operationFailedWithMessage unexpected error: $errorDetails",
```

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/enumerator/ApplicationExceptionCodeEnum.kt [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package com.nosql.authservice.common.exception.enumerator
+
+enum class ApplicationExceptionCodeEnum(val message: String) {
+    DATA_INTEGRITY_VIOLATION_ERROR("Data integrity violation error"),
+    DECODING_ERROR("Invalid request"),
+    NOT_FOUND_ERROR("Not found any record by provided parameters"),
+    UNEXPECTED_ERROR("Unexpected application error"),
+    USER_ALREADY_EXISTS_ERROR("User with given id already exists"),
+    VALIDATION_ERROR("Invalid data"),
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/enumerator/ApplicationExceptionCodeEnum.kt]: error -> exception

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt (hunk 2 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt (hunk 3 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt (hunk 4 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt (hunk 5 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/handler/ExceptionHandler.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 9 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+        return when (error) {
+            is DataIntegrityViolationException -> {
+                DataIntegrityViolationApplicationException()
+            }
+            is ApplicationException -> {
+                error.apply { description = """$operationFailedWithMessage error: $errorDetails""" }
+            }
+            else -> {
+                ApplicationException(
+                    code = ApplicationExceptionCodeEnum.UNEXPECTED_ERROR,
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: I think you should make `UNEXPECTED_ERROR` as a default value for the `ApplicationException.code` field

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 10 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 11 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 12 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/configuration/MongoConfig.kt [hunk 18]
```diff
@@ -8,13 +8,13 @@ import org.springframework.data.mongodb.config.AbstractReactiveMongoConfiguratio
 import org.springframework.data.mongodb.repository.config.EnableReactiveMongoRepositories
 
 @Configuration
-@EnableReactiveMongoRepositories(basePackageClasses = [MongoConfig::class])
+@EnableReactiveMongoRepositories(basePackages = ["com.nosql.authservice.repository"])
 class MongoConfig : AbstractReactiveMongoConfiguration() {
 
     @Value("\${spring.data.mongodb.uri}")
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/configuration/MongoConfig.kt]: Could you use properties class?

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 13 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

**Review by V7074 — CHANGES_REQUESTED** [review]

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 14 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 15 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/controller/UserController.kt [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    suspend fun signUp(
+        @Valid @RequestBody userDto: UserDto,
+    ) = userService.signUp(userDto)
+
+    @PostMapping(
+        path = [SING_IN_URL_PATH],
+        consumes = [MediaType.APPLICATION_JSON_VALUE],
+        produces = [MediaType.APPLICATION_JSON_VALUE],
+    )
+    suspend fun save(
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/controller/UserController.kt]: save?

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/enumerator/ApplicationExceptionCodeEnum.kt (hunk 16 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/common/exception/enumerator/ApplicationExceptionCodeEnum.kt]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt (hunk 17 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt]: fixed

**Review by V7074 — CHANGES_REQUESTED** [review]

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt [hunk 20]
```diff
@@ -0,0 +1,6 @@
+package com.nosql.authservice.dto
+
+data class SignInResponseDto (
+
+    var id: String,
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt]: val

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt [hunk 21]
```diff
@@ -0,0 +1,6 @@
+package com.nosql.authservice.dto
+
+data class SignInResponseDto (
+
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt:6]: may be add `@JsonProperty` annotation here?

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/UserDto.kt [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+
+    @JsonProperty("login")
+    @field:NotBlank(message = "login must not be null or blank")
+    @field:Size(min = 1, max = 15, message = "login size must be from 1 to 15")
+    @field:Pattern(regexp = "^[a-z0-9_-]{1,15}\$", message = "login contains unacceptable symbols")
+    val login: String,
+
+    @JsonProperty("password_hash")
+    @field:NotBlank(message = "password must not be null or blank")
+    @field:Size(min = 1, max = 32, message = "password size must be from 1 to 32")
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/UserDto.kt:22]: pattern?

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/entity/UserEntity.kt [hunk 23]
```diff
@@ -0,0 +1,19 @@
+package com.nosql.authservice.entity
+
+import org.bson.types.ObjectId
+import org.springframework.data.annotation.Id
+import org.springframework.data.mongodb.core.mapping.Document
+import org.springframework.data.mongodb.core.mapping.Field
+
+@Document(collection = "auth")
```
Comment (V7074) [inline backend/auth-service/src/main/kotlin/com/nosql/authservice/entity/UserEntity.kt]: auth -> user
or something else

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/configuration/MongoConfig.kt (hunk 18 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/configuration/MongoConfig.kt]: Will be correct later

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/controller/UserController.kt (hunk 19 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/controller/UserController.kt]: fixed

File: backend/auth-service/src/main/resources/application-local.yaml [hunk 24]
```diff
@@ -0,0 +1,4 @@
+spring:
+  data:
+    mongodb:
+      uri: mongodb://mongo:mongo@localhost:27017/it-service-db?authSource=admin
```
Comment (V7074) [inline backend/auth-service/src/main/resources/application-local.yaml]: change the db to auth specific, like this: `auth`

**Review by V7074 — CHANGES_REQUESTED** [review]

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/entity/UserEntity.kt (hunk 23 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/entity/UserEntity.kt]: fixed

File: backend/auth-service/src/main/resources/application-local.yaml (hunk 24 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/resources/application-local.yaml]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/UserDto.kt (hunk 22 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/UserDto.kt:22]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt (hunk 21 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt:6]: fixed

File: backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt (hunk 20 above)
Comment (AksyonovaKaterina) [reply backend/auth-service/src/main/kotlin/com/nosql/authservice/dto/SignInResponseDto.kt]: fixed

**Review by V7074 — APPROVED** [review]

**Pull request merged**